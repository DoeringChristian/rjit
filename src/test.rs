use crate::Trace;
use anyhow::*;

macro_rules! test_uop {
        ($jop:ident($init:expr; $ty:ident) $(,$mod:literal)?) => {
            test_uop!($ty::$jop => $jop($init; $ty) $(,$mod)?);
        };
        ($rop:expr => $jop:ident($init:expr; $ty:ident) $(,$mod:literal)?) => {
            paste::paste! {
                #[test]
                fn [<$jop _$ty $(__$mod)?>]() -> Result<()> {
                    // pretty_env_logger::init();

                    let initial: &[$ty] = &$init;

                    let ir = Trace::default();
                    ir.set_backend(["cuda"])?;

                    let x = ir.array(initial)?;

                    let y = x.$jop()?;

                    ir.schedule(&[&y]);

                    ir.eval()?;

                    insta::assert_snapshot!(ir.kernel_history());

                    for (i, calculated) in y.to_host::<$ty>()?.into_iter().enumerate(){
                        let expected = ($rop)(initial[i]);
                        approx::assert_abs_diff_eq!(calculated, expected, epsilon = 0.0001);
                    }
                    Ok(())
                }
            }
        };
    }

test_uop!(|x:f32| {1./x} => rcp(         [0.1, 0.5, 1., std::f32::consts::PI]; f32));
test_uop!(|x:f32| {1./x.sqrt()} => rsqrt([0.1, 0.5, 1., std::f32::consts::PI]; f32));
test_uop!(sin(                           [0.1, 0.5, 1., std::f32::consts::PI]; f32));
test_uop!(cos(                           [0.1, 0.5, 1., std::f32::consts::PI]; f32));
test_uop!(exp2(                          [0.1, 0.5, 1., std::f32::consts::PI]; f32));
test_uop!(log2(                          [0.1, 0.5, 1., std::f32::consts::PI]; f32));

#[test]
fn opaque() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["cuda"])?;

    let x = ir.literal(10u32)?;
    let x = x.opaque();

    let y = ir.array(&[1u32, 2, 3])?;

    let y = y.add(&x)?;

    y.schedule();
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(y.to_host::<u32>()?, vec![11u32, 12, 13]);
    Ok(())
}
#[test]
fn make_opaque() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["cuda"])?;

    let x = ir.literal(10u32)?;
    x.make_opaque();

    let y = ir.array(&[1u32, 2, 3])?;

    let y = y.add(&x)?;

    y.schedule();
    ir.eval()?;

    insta::assert_snapshot!(ir.kernel_history());

    assert_eq!(y.to_host::<u32>().unwrap(), vec![11, 12, 13]);
    Ok(())
}

#[test]
fn compress_small() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["cuda"])?;

    // let x = ir.array(&[true, false, true]);
    let x = ir.array(&[true, false, true, false, true, false, true, false])?;

    let i = x.compress()?;

    assert_eq!(i.to_host::<u32>().unwrap(), vec![0, 2, 4, 6]);
    Ok(())
}
#[test]
fn compress_large() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["cuda"])?;

    const N: u32 = 4100;

    let x = ir.array(&vec![true; N as usize])?;

    let i = x.compress()?;

    assert_eq!(i.to_host::<u32>().unwrap(), (0..N).collect::<Vec<_>>());
    Ok(())
}
#[test]
fn compress_small_optix() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    // let x = ir.array(&[true, false, true]);
    let x = ir.array(&[true, false, true, false, true, false, true, false])?;

    let i = x.compress()?;

    assert_eq!(i.to_host::<u32>().unwrap(), vec![0, 2, 4, 6]);
    Ok(())
}
#[test]
fn compress_large_optix() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    const N: u32 = 4100;

    let x = ir.array(&vec![true; N as usize])?;

    let i = x.compress()?;

    assert_eq!(i.to_host::<u32>().unwrap(), (0..N).collect::<Vec<_>>());
    Ok(())
}
#[test]
fn kernel_reuse() -> Result<()> {
    let ir = Trace::default();
    ir.set_backend(["optix"])?;

    for _ in 0..100 {
        let mut x = ir.array(&[0, 1, 2, 3, 4])?;
        for j in 0..100 {
            x = x.add(&ir.literal(j)?)?;
        }

        x.schedule();
        ir.eval()?;
    }

    println!("{}", ir.kernel_history());
    assert_eq!(ir.kernel_cache_size(), 1);

    Ok(())
}
