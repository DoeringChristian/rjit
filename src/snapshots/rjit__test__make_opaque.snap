---
source: src/test.rs
expression: ir.kernel_history()
---
Kernel History:
Launched CUDA Kernel 8364512272315240869 with 3 elements
===============================================
Kernel 8364512272315240869:

.version 8.0
.target sm_86
.address_size 64

.entry cujit(
	.param .align 8 .b8 params[40]) {

	.reg.b8   %b <14>; .reg.b16 %w<14>; .reg.b32 %r<14>;
	.reg.b64  %rd<14>; .reg.f32 %f<14>; .reg.f64 %d<14>;
	.reg.pred %p <14>;

	mov.u32 %r0, %ctaid.x;
	mov.u32 %r1, %ntid.x;
	mov.u32 %r2, %tid.x;
	mad.lo.u32 %r0, %r0, %r1, %r2; // r0 <- Index

	// Index Conditional (jump to done if Index >= Size).
	ld.param.u32 %r2, [params]; // r2 <- params[0] (Size)
	setp.ge.u32 %p0, %r0, %r2; // p0 <- r0 >= r2
	@%p0 bra done; // if p0 => done
	
	mov.u32 %r3, %nctaid.x; // r3 <- nctaid.x
	mul.lo.u32 %r1, %r3, %r1; // r1 <- r3 * r1
	
body: // sm_86

	// [0]: ScheduleVar { op: Data, deps: [], ty: U32, reg: 5, data: Buffer(0), sbt_hash: 0 } =>

	// [1]: ScheduleVar { op: Idx, deps: [], ty: U32, reg: 6, data: None, sbt_hash: 0 } =>
	mov.u32 %r6, %r0;


	// [2]: ScheduleVar { op: Literal, deps: [], ty: Bool, reg: 7, data: Literal(1), sbt_hash: 0 } =>
	mov.pred %p7, 0x1;


	// [3]: ScheduleVar { op: Gather, deps: [SVarId(0), SVarId(1), SVarId(2)], ty: U32, reg: 8, data: Buffer(1), sbt_hash: 0 } =>
	ld.param.u64 %rd0, [params+16];
	mad.wide.u32 %rd3, %r6, 4, %rd0;
	ld.global.nc.u32 %r8, [%rd3];

	// [4]: ScheduleVar { op: Literal, deps: [], ty: U32, reg: 9, data: Opaque(0), sbt_hash: 0 } =>
	ld.param.b32 %r9, [params+8];

	// [5]: ScheduleVar { op: Add, deps: [SVarId(3), SVarId(4)], ty: U32, reg: 10, data: Buffer(2), sbt_hash: 0 } =>
	add.u32 %r10, %r8, %r9;

	// [6]: ScheduleVar { op: Idx, deps: [], ty: U32, reg: 11, data: None, sbt_hash: 0 } =>
	mov.u32 %r11, %r0;


	// [7]: ScheduleVar { op: Literal, deps: [], ty: Bool, reg: 12, data: Literal(1), sbt_hash: 0 } =>
	mov.pred %p12, 0x1;


	// [8]: ScheduleVar { op: Scatter { op: None }, deps: [SVarId(5), SVarId(5), SVarId(6), SVarId(7)], ty: U32, reg: 13, data: None, sbt_hash: 0 } =>
	ld.param.u64 %rd0, [params+32];
	mad.wide.u32 %rd3, %r11, 4, %rd0;
	st.global.u32 [%rd3], %r10;

	//End of Kernel:

	add.u32 %r0, %r0, %r1; // r0 <- r0 + r1
	setp.ge.u32 %p0, %r0, %r2; // p0 <- r1 >= r2
	@!%p0 bra body; // if p0 => body


done:

	ret;
}

